lpr_daemon.go

```diff
@@ -402,213 +472,231 @@ func (lpr *LprConnection) Close() {
-			lpr.Output = nil
-		}
-		lpr.tempFilesize = lpr.tempFilesize - length
-		lpr.Status = PRINTJOB_SUB_COMMANDS
-	}
-
-	// pro := float32(100.0) - float32(lpr.tempFilesize*100)/float32(lpr.Filesize)
-	// fmt.Print("<")
-	// for i := 0; i < 50; i++ {
-	// 	if int(pro/2) < i {
-	// 		fmt.Print(" ")
-	// 	} else {
-	// 		fmt.Print("-")
-	// 	}
-	// }
-	// fmt.Printf("> %f %%\r", pro)
-}
-
-// Interpret interprets the LPR daemon commands
-func (lpr *LprConnection) Interpret(data []uint8, length int64) {
-	firstSymbol := data[0]
 	switch firstSymbol {
 	/* Daemon commands */
 	/* 01 - Print any waiting jobs */
 	case 0x1:
+		lpr.typeChan <- ConnectionTypePrintAnyWaitingJobs
 
 	/* 02 - Receive a printer job */
 	case 0x2:
+		lpr.typeChan <- ConnectionTypeReceivePrintJob
 		var err error
-		lpr.PrqName, _, err = lpr.daemon.ensureUTF8(data[1:length])
+		lpr.PrqName, _, err = lpr.daemon.ensureUTF8(command[1:])
 		if err != nil {
 			logErrorf("Invalid printer queue name %q: %v", lpr.PrqName, err)
 		}
-		lpr.Status = PRINTJOB_SUB_COMMANDS
+		lpr.Status = JobSubCommand
+
+		return lpr.sendAck()
 
 	/* 03 - Send queue state (short) */
 	/* | 03 | Queue | SP | List | LF | */
 	case 0x3:
-		fallthrough
+		lpr.typeChan <- ConnectionTypeSendQueueStateShort
+		return lpr.sendQueueState(command, false)
 
 	/* 04 - Send queue state (long) */
 	/* | 04 | Queue | SP | List | LF | */
 	case 0x4:
-		content := string(data[1:length])
-		parts := strings.SplitN(content, " ", 2)
-		queue := parts[0]
-		list := ""
-		if len(parts) > 1 {
-			list = parts[1]
-		}
-
-		lpr.replyQueueState(queue, list, firstSymbol == 0x4)
+		lpr.typeChan <- ConnectionTypeSendQueueStateLong
+		return lpr.sendQueueState(command, true)
 
 	/* 05 - Remove jobs */
 	case 0x5:
+		lpr.typeChan <- ConnectionTypeRemoveJobs
 
 	default:
-		logErrorf("First Element: %02x (%c)", data[0], data[0])
-		logErrorf("Unknown Code: %s", string(data[:length]))
-		break
+		lpr.typeChan <- ConnectionTypeUnknown
+		return fmt.Errorf("unknown Daemon command %02x (%c) :: %s", command[0], command[0], string(command))
+
 	}
+
+	return nil
 }
 
-func (lpr *LprConnection) createTempFile() (*os.File, error) {
-	try := 0
-	for {
-		fileName := filepath.Join(lpr.daemon.InputFileSaveDir, strconv.FormatUint(uint64(rand.Int63()), 16))
+func (lpr *LprConnection) sendQueueState(command []byte, long bool) error {
+	parts := operands(command[1:], 2)
+	queue := parts[0]
+	list := ""
+	if len(parts) > 1 {
+		list = parts[1]
+	}
 
-		f, err := os.OpenFile(fileName, os.O_RDWR|os.O_CREATE|os.O_EXCL, lpr.daemon.fileMask)
-		if os.IsExist(err) {
-			if try++; try < 10000 {
-				continue
+	return lpr.replyQueueState(queue, list, long)
+}
+
+var asciiSpace = [256]byte{' ': 1, '\t': 1, '\v': 1, '\f': 1}
+
+func operands(data []byte, max int) []string {
+	var opers = make([]string, 0)
+
+	oper := []byte{}
+	for i, b := range data {
+		if asciiSpace[b] == 1 {
+			opers = append(opers, string(oper))
+
+			if len(opers) == max {
```

Comment:
I don't think this check is correct. I think a queue state command like `| 03 | raw | SP | listentry1 listentry2 | LF |` would cause an incorrect list (`listentry1` instead of `listentry1 listentry2`)